File: src/main/java/com/aiproxyoauth/auth/AuthManager.java
package com.aiproxyoauth.auth;

import com.fasterxml.jackson.databind.JsonNode;
import com.aiproxyoauth.config.ServerConfig;
import com.aiproxyoauth.model.CodexClientVersionResolver;
import com.aiproxyoauth.util.JwtParser;

import java.net.http.HttpClient;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

public class AuthManager implements CredentialsProvider {

    private static final long REFRESH_EXPIRY_MARGIN_MS = 5 * 60 * 1000L;

    private final ServerConfig config;
    private final HttpClient httpClient;
    private final ReentrantLock lock = new ReentrantLock();
    private volatile AuthLoader.AuthResult current;

    public AuthManager(ServerConfig config, HttpClient httpClient) {
        this.config = config;
        this.httpClient = httpClient;
    }

    public AuthLoader.AuthResult ensureFresh() throws Exception {
        lock.lock();
        try {
            // Re-check after acquiring the lock: another thread may have already refreshed.
            AuthLoader.AuthResult existing = current;
            if (existing != null && !isTokenExpiringSoon(existing.accessToken())) {
                return existing;
            }
            current = AuthLoader.loadAuthTokens(
                    config.oauthFilePath(),
                    config.oauthClientId(),
                    null, // issuer derived from defaults
                    config.oauthTokenUrl(),
                    httpClient
            );
            return current;
        } finally {
            lock.unlock();
        }
    }

    /**
     * Drops the cached token and reloads from the auth file. Returns true only when the
     * reload produced a different access token than the rejected one — reloading the same
     * locally-valid-but-upstream-rejected token would make a retry pointless.
     */
    @Override
    public boolean refreshAfterUnauthorized(String rejectedAuthorizationHeader) throws Exception {
        lock.lock();
        try {
            current = null;
        } finally {
            lock.unlock();
        }
        AuthLoader.AuthResult refreshed = ensureFresh();
        if (refreshed == null || refreshed.accessToken() == null || refreshed.accessToken().isEmpty()) {
            return false;
        }
        String rejectedToken = stripBearer(rejectedAuthorizationHeader);
        return !refreshed.accessToken().equals(rejectedToken);
    }

    private static String stripBearer(String authorizationHeader) {
        if (authorizationHeader == null) {
            return null;
        }
        String value = authorizationHeader.strip();
        return value.regionMatches(true, 0, "Bearer ", 0, 7) ? value.substring(7).strip() : value;
    }

    @Override
    public Map<String, String> getAuthHeaders() throws Exception {
        AuthLoader.AuthResult auth = current;
        if (auth == null || isTokenExpiringSoon(auth.accessToken())) {
            auth = ensureFresh();
        }
        // Map.of() rejects null values with NullPointerException. accountId() is safe here
        // because AuthLoader.loadAuthTokens() throws IOException before returning an AuthResult
        // with a null or empty accountId, so ensureFresh() would have propagated that exception
        // before we reach this point.
        return Map.of(
                "Authorization", "Bearer " + auth.accessToken(),
                "chatgpt-account-id", auth.accountId(),
                "version", CodexClientVersionResolver.resolve(config.codexVersion())
        );
    }

    private static boolean isTokenExpiringSoon(String accessToken) {
        if (accessToken == null || accessToken.isEmpty()) return true;
        JsonNode claims = JwtParser.parseClaims(accessToken);
        if (claims != null && claims.has("exp") && claims.get("exp").isNumber()) {
            long expiryMs = claims.get("exp").asLong() * 1000;
            return expiryMs <= System.currentTimeMillis() + REFRESH_EXPIRY_MARGIN_MS;
        }
        return false;
    }

    public AuthLoader.AuthResult getCurrent() {
        return current;
    }
}

File: src/main/kotlin/de/moritzf/proxy/auth/AuthManager.kt
package de.moritzf.proxy.auth

import de.moritzf.proxy.config.ServerConfig
import de.moritzf.proxy.model.CodexClientVersionResolver
import de.moritzf.proxy.util.JwtParser
import com.fasterxml.jackson.databind.JsonNode
import java.net.http.HttpClient
import java.util.concurrent.locks.ReentrantLock

class AuthManager(
    private val config: ServerConfig,
    private val httpClient: HttpClient,
) : CredentialsProvider {
    private val lock = ReentrantLock()

    @Volatile
    private var current: AuthLoader.AuthResult? = null

    @Throws(Exception::class)
    fun ensureFresh(): AuthLoader.AuthResult {
        lock.lock()
        try {
            // Re-check after acquiring the lock: another thread may have already refreshed.
            val existing = current
            if (existing != null && !isTokenExpiringSoon(existing.accessToken())) {
                return existing
            }
            val loaded = AuthLoader.loadAuthTokens(
                config.oauthFilePath(),
                config.oauthClientId(),
                null, // issuer derived from defaults
                config.oauthTokenUrl(),
                httpClient,
            )
            current = loaded
            return loaded
        } finally {
            lock.unlock()
        }
    }

    /**
     * Drops the cached token and reloads from the auth file. Returns true only when the
     * reload produced a different access token than the rejected one: reloading the same
     * locally-valid-but-upstream-rejected token would make a retry pointless.
     */
    @Throws(Exception::class)
    override fun refreshAfterUnauthorized(rejectedAuthorizationHeader: String?): Boolean {
        lock.lock()
        try {
            current = null
        } finally {
            lock.unlock()
        }
        val refreshed = ensureFresh()
        if (refreshed.accessToken().isEmpty()) {
            return false
        }
        val rejectedToken = stripBearer(rejectedAuthorizationHeader)
        return refreshed.accessToken() != rejectedToken
    }

    @Throws(Exception::class)
    override fun getAuthHeaders(): Map<String, String> {
        var auth = current
        if (auth == null || isTokenExpiringSoon(auth.accessToken())) {
            auth = ensureFresh()
        }
        // Java Map.of() rejected null values. accountId() is safe here because
        // AuthLoader.loadAuthTokens() throws before returning a missing account id.
        return mapOf(
            "Authorization" to "Bearer ${auth.accessToken()}",
            "chatgpt-account-id" to auth.accountId(),
            "version" to CodexClientVersionResolver.resolve(config.codexVersion()),
        )
    }

    @Suppress("unused")
    fun getCurrent(): AuthLoader.AuthResult? = current

    companion object {
        private const val REFRESH_EXPIRY_MARGIN_MS = 5 * 60 * 1000L

        private fun stripBearer(authorizationHeader: String?): String? {
            if (authorizationHeader == null) {
                return null
            }
            val value = authorizationHeader.trim()
            return if (value.startsWith("Bearer ", ignoreCase = true)) value.substring(7).trim() else value
        }

        private fun isTokenExpiringSoon(accessToken: String?): Boolean {
            if (accessToken.isNullOrEmpty()) {
                return true
            }
            val claims: JsonNode? = JwtParser.parseClaims(accessToken)
            if (claims != null && claims.has("exp") && claims.get("exp").isNumber) {
                val expiryMs = claims.get("exp").asLong() * 1000
                return expiryMs <= System.currentTimeMillis() + REFRESH_EXPIRY_MARGIN_MS
            }
            return false
        }
    }
}
